neumai-tools/neumai_tools/PipelineCollection/PipelineCollection.py

```diff
@@ -53,4 +54,29 @@ def search_separate(self, query:str, number_of_results:int)-> List:
     def search_routed(self, query:str, number_of_results:int)-> List:
         """Routed search based on the contents available in a pipeline"""
         # Need to add descriptions to the pipeline and generate a basic index on top of them
-        raise NotImplementedError("In the works. Contact founders@tryneum.com for information")
\ No newline at end of file
+
+        pipe_to_similarity = {}
+        for pipe in self.pipelines:
+            pipe_representative = pipe.sink.get_representative_vector()
+            query_vector = pipe.embed.embed_query(query=query)
+            distance_from_representative = cosine(pipe_representative, query_vector)
+
+            # Similarity score, hence subtracted distance from 1
+            pipe_to_similarity[pipe.id] = 1 - distance_from_representative
+
+        # We want to sort by decreasing oeder of similarity score
+        # The more similar the query to a given representative vector
+        # the higher rank that pipeline would get in terms of search.
+        # Currently, we are only selection the only pipeline whose 
+        # representative is most similar to the query.
+        pipe_to_similarity = dict(sorted(pipe_to_similarity.items(), key=lambda x: x[1], reverse=True)[:1])
```

Comment:
Can probably make the number of pipelines to use configurable. Assume in most cases you might want to pick the highest. Also might want to consider making it configurable to using any pipelines with a similarity score above X.

Reply:
The more I think about it, I think that making it configurable to similarities above a certain threshold seems best.

Reply:
@ddematheu Yep, sounds good. But deciding that threshold is difficult. I mean, you never know even the most similar pipe can have similarity scores that are low sounding like 0.4 etc. We need some way to decide this threshold as per the computed similarities.

Reply:
100% agree with this.maybe keeping to top to start is fine. My thinking is that we could provide different modalities. One where we pass results for N highest pipelines (default to N=1) and another for threshold based where we pass results for pipelines with similarity is higher than N (default N = 0.5 or whatever)

One additional modality might be to look at distribution. Where N is the max distance or % away from the highest scoring.

Btw this are just thoughts, happy to commit to starting with highest and add other modalities based on feedback.

Reply:
These are interesting ideas. Especially the distribution one, sounds similar to p-sampling in language models.